src/py/flwr/server/compat/app_utils.py

```diff
@@ -0,0 +1,100 @@
+# Copyright 2024 Flower Labs GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Utility functions for the `start_driver`."""
+
+
+import threading
+import time
+from typing import Dict, Tuple
+
+from ..client_manager import ClientManager
+from ..compat.driver_client_proxy import DriverClientProxy
+from ..driver import Driver
+
+
+def start_update_client_manager_thread(
+    driver: Driver,
+    client_manager: ClientManager,
+) -> Tuple[threading.Thread, threading.Event]:
+    """Periodically update the nodes list in the client manager in a thread.
+
+    This function starts a thread that periodically uses the associated driver to
+    get all node_ids. Each node_id is then converted into a `DriverClientProxy`
+    instance and stored in the `registered_nodes` dictionary with node_id as key.
+
+    New nodes will be added to the ClientManager via `client_manager.register()`,
+    and dead nodes will be removed from the ClientManager via
+    `client_manager.unregister()`.
+
+    Parameters
+    ----------
+    driver : Driver
+        The Driver object to use.
+    client_manager :ClientManager
+        The ClientManager object to be updated.
+
+    Returns
+    -------
+    Tuple[threading.Thread, threading.Event]
+        The thread updating the ClientManager and the stop event.
+    """
```

Comment:
The `threading.Event()` needs to be added here

Reply:
Updated